features/util.py

```diff
@@ -85,72 +85,87 @@ def lxc_get_property(name: str, property_name: str, image: bool = False):
-    :param cloud_api: Optional pycloudlib BaseCloud api if available for the
-        machine_type
+    :param cache_source: If False, we will always rebuild the source environemt
+                         for sbuild
 
     :return: A list of file paths to debs created by the build.
-
     """
-    if os.environ.get("TRAVIS") != "true":
-        logging.info(
-            "--- Assuming non-travis build. Creating: {}".format(SOURCE_PR_TGZ)
+    logging.info("--- Creating: {}".format(SOURCE_PR_TGZ))
+    temp_dir = tempfile.gettempdir()
+    if not os.path.exists(os.path.dirname(SOURCE_PR_TGZ)):
+        os.makedirs(os.path.dirname(SOURCE_PR_TGZ))
+
+    if not os.path.exists(SOURCE_PR_TGZ) or not cache_source:
+        cwd = os.getcwd()
+        os.chdir("..")
+        subprocess.run(
+            [
+                "tar",
+                "-zcf",
+                SOURCE_PR_TGZ,
+                "--exclude-vcs",
+                "--exclude-vcs-ignores",
+                os.path.basename(cwd),
+            ]
         )
-        if not os.path.exists(os.path.dirname(SOURCE_PR_TGZ)):
-            os.makedirs(os.path.dirname(SOURCE_PR_TGZ))
-        if not os.path.exists(SOURCE_PR_TGZ) or not cache_source:
-            cwd = os.getcwd()
-            os.chdir("..")
-            subprocess.run(
-                [
-                    "tar",
-                    "-zcf",
-                    SOURCE_PR_TGZ,
-                    "--exclude-vcs",
-                    "--exclude-vcs-ignores",
-                    os.path.basename(cwd),
-                ]
-            )
-            os.chdir(cwd)
-    buildscript = "build-from-source.sh"
-    with open(buildscript, "w") as stream:
-        stream.write(BUILD_FROM_TGZ)
-
-    instance = cloud_api.get_instance(instance_id=container_name)
-    for filepath in (buildscript, SOURCE_PR_TGZ):
-        logging.info("--- Push {} -> {}:/tmp".format(filepath, instance.name))
-        instance.push_file(filepath, "/tmp/" + os.path.basename(filepath))
-    result = instance.execute(["sudo", "bash", "/tmp/" + buildscript])
-    if result.failed:
-        logging.error("--- ERROR: Building PR failed")
-        raise Exception(
-            "stdout: {}\nstderr: {}".format(result.stdout, result.stderr)
+        os.chdir(cwd)
+
+    logging.info("--- Creating: {}".format(SOURCE_PR_UNTAR_DIR))
+    if not os.path.exists(SOURCE_PR_UNTAR_DIR) or not cache_source:
+        os.makedirs(SOURCE_PR_UNTAR_DIR)
+        subprocess.run(
+            [
+                "tar",
+                "-xvf",
+                SOURCE_PR_TGZ,
+                "-C",
+                SOURCE_PR_UNTAR_DIR,
+                "--strip-components=1",
+            ],
+            stdout=subprocess.DEVNULL,
+            stderr=subprocess.DEVNULL,
         )
-    deb_artifacts = []
+
+    curr_dir = os.getcwd()
+    os.chdir(temp_dir)
+    logging.info("--- Running sbuild")
+    subprocess.run(
+        [
+            "sbuild",
+            "--no-run-lintian",
```

Comment:
We could speed up the process more by adding the flags `--no-apt-upgrades` and `--no-apt-distupgrades`. However, I don't know if this a good approach for building on the chroots

Reply:
In general it's good we want the chroot to be up-to-date, otherwise when installing the build-deps there may be packages in the local package index that are not present anymore in the archive (due to updated). However we can assume that *on the jenkins nodes* the schroots are up-to-date, because that's done automatically every time a schroot session is started (that's how `sbuild-launchpad-chroot` sets things up).

Reply:
Given that the schroot will be up to date the performance penalty will be rather low, I'd stick with the defaults to cover more use cases.